src/Orleans.Core/Messaging/ClientMessageCenter.cs

```diff
@@ -103,16 +106,66 @@ internal class ClientMessageCenter : IMessageCenter, IDisposable
             }
         }
 
-        public void Start()
+        public async Task StartAsync(CancellationToken cancellationToken)
         {
-            Running = true;
+            await EstablishInitialConnection(cancellationToken);
+
             if (this.statisticsLevel.CollectQueueStats())
             {
                 queueTracking.OnStartExecution();
             }
+
+            Running = true;
             if (logger.IsEnabled(LogLevel.Debug)) logger.Debug("Proxy grain client started");
         }
 
+        private async Task EstablishInitialConnection(CancellationToken cancellationToken)
+        {
+            var cancellationTask = cancellationToken.WhenCancelled();
+            var liveGateways = gatewayManager.GetLiveGateways();
+            var pendingTasks = new List<Task>(liveGateways.Count + 1);
+            pendingTasks.Add(cancellationTask);
+            foreach (var gateway in liveGateways)
+            {
+                pendingTasks.Add(connectionManager.GetConnection(gateway).AsTask());
+            }
+
+            if (liveGateways.Count == 0)
```

Comment:
Can we move this check to before we populate `pendingTasks`?

Reply:
Yes, good point